### PR TITLE
chore: allow opensea price discovery in core-sdk

### PR DESCRIPTION
**Pull request by levalleux-ludo**
## Description

{{what was done}}

## How to test

{{how can it be tested}}

### CODE REVIEW

File: e2e/opensea-api-mock/src/services/opensea.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    } else if (!nftBid && nftAsk) {
+      side = OrderSide.ASK;
+    } else if (nftBid && nftAsk) {
+      logger.warn(
+        `NFT found in both consideration and offer. Unable to detect the order side`
+      );
+    } else {
+      logger.warn(
+        `No NFT found in consideration or offer. Unable to detect the order side`
+      );
```
Comment (albertfolch-redeemeum) [inline e2e/opensea-api-mock/src/services/opensea.ts:172]: why not throwing an error?

File: e2e/opensea-api-mock/src/utils/logger.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import winston, { format } from "winston";
```
Comment (albertfolch-redeemeum) [inline e2e/opensea-api-mock/src/utils/logger.ts:1]: why both winston and morgan? and not just one?

File: e2e/tests/utils.ts [hunk 3]
```diff
@@ -1353,3 +1363,44 @@ export function serializeVariant(variant: productV1.ProductV1Variant): string {
   );
   return JSON.stringify(orderedTable).toLowerCase();
 }
+
+export function createOpenseaSdk(privateKey: string): OpenSeaSDK {
+  const providerV6 = new JsonRpcProviderV6(defaultConfig.jsonRpcUrl);
+  const walletV6: WalletV6 = new WalletV6(privateKey, providerV6);
+  const OPENSEA_API_KEY = ""; // local mock does'nt need any
```
Comment (albertfolch-redeemeum) [inline e2e/tests/utils.ts:1370]:
```suggestion
  const OPENSEA_API_KEY = ""; // local mock doesn't need any
```

File: packages/common/src/types/offers.ts [hunk 4]
```diff
@@ -73,3 +73,17 @@ export type PremintParametersStruct = {
   reservedRangeLength: BigNumberish;
   to: string;
 };
+
+export enum Side {
+  Ask = 0,
+  Bid = 1,
+  Wrapper = 2
```
Comment (albertfolch-redeemeum) [inline packages/common/src/types/offers.ts:80]: if possible, I'd not use 0
```suggestion
  Ask = 1,
  Bid = 2,
  Wrapper = 3
```

File: packages/core-sdk/src/seaport/interface.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+export enum eOrderType {
+  FULL_OPEN = 0,
+  PARTIAL_OPEN = 1,
+  FULL_RESTRICTED = 2,
+  PARTIAL_RESTRICTED = 3,
+  CONTRACT = 4
+}
+
+export enum eItemType {
+  NATIVE = 0,
```
Comment (albertfolch-redeemeum) [inline packages/core-sdk/src/seaport/interface.ts:16]: let's try to not use the value 0 if possible and if it makes sense

File: scripts/create-price-discovery-offer.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    rawData.toString()
+  ) as offers.CreateOfferArgs;
+  if (offerDataJson["validFromDateInMS"] === undefined) {
+    offerDataJson["validFromDateInMS"] = Date.now();
+  }
+  if (offerDataJson["voucherRedeemableFromDateInMS"] === undefined) {
+    offerDataJson["voucherRedeemableFromDateInMS"] = Date.now();
+  }
+  // const WETH_ADDRESS = "0x52eF3d68BaB452a294342DC3e5f464d7f610f72E"; // Amoy
+  const WETH_ADDRESS = "0x7b79995e5f793a07bc00c21412e50ecae098e7f9"; // Sepolia
```
Comment (albertfolch-redeemeum) [inline scripts/create-price-discovery-offer.ts:39]: use one or the other depending on configId/envName?

File: scripts/opensea/create-listing.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  const provider = new providers.JsonRpcProvider(defaultConfig.jsonRpcUrl);
+  const providerV6 = new JsonRpcProviderV6(defaultConfig.jsonRpcUrl);
+  const sellerWallet = new Wallet(sellerPrivateKey);
+  const sellerWalletV6: WalletV6 = new WalletV6(sellerPrivateKey, providerV6);
+  const coreSDK = CoreSDK.fromDefaultConfig({
+    web3Lib: new EthersAdapter(provider, sellerWallet),
+    envName,
+    configId
+  });
+  const WETH_ADDRESS = "0xA6FA4fB5f76172d178d61B04b0ecd319C5d1C0aa";
```
Comment (albertfolch-redeemeum) [inline scripts/opensea/create-listing.ts:43]: which chain?

File: scripts/opensea/create-listing.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+  const { offerId, exchangeId } = coreSDK.parseTokenId(tokenId);
+  console.log({
+    offerId: offerId.toString(),
+    exchangeId: exchangeId.toString()
+  });
+  const openseaUrl = API_BASE_TESTNET;
+  // const openseaUrl = "http://localhost:3334";
```
Comment (albertfolch-redeemeum) [inline scripts/opensea/create-listing.ts:51]: remove?

File: scripts/opensea/create-offer-opensea.ts [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  const provider = new providers.JsonRpcProvider(defaultConfig.jsonRpcUrl);
+  const providerV6 = new JsonRpcProviderV6(defaultConfig.jsonRpcUrl);
+  const buyerWallet = new Wallet(buyerPrivateKey);
+  const sellerWalletV6: WalletV6 = new WalletV6(buyerPrivateKey, providerV6);
+  const coreSDK = CoreSDK.fromDefaultConfig({
+    web3Lib: new EthersAdapter(provider, buyerWallet),
+    envName,
+    configId
+  });
+  const WETH_ADDRESS = "0xA6FA4fB5f76172d178d61B04b0ecd319C5d1C0aa";
```
Comment (albertfolch-redeemeum) [inline scripts/opensea/create-offer-opensea.ts:41]: which chain?

**Review by albertfolch-redeemeum — APPROVED** [review]

File: scripts/opensea/create-offer-opensea.ts (hunk 9 above)
Comment (levalleux-ludo) [reply scripts/opensea/create-offer-opensea.ts:41]: Changed. Now reading the WETH address from the coreSDK config

File: scripts/opensea/create-listing.ts (hunk 8 above)
Comment (levalleux-ludo) [reply scripts/opensea/create-listing.ts:51]: Changed

File: scripts/opensea/create-listing.ts (hunk 7 above)
Comment (levalleux-ludo) [reply scripts/opensea/create-listing.ts:43]: Changed. Now reading the WETH address from the coreSDK config

File: scripts/create-price-discovery-offer.ts (hunk 6 above)
Comment (levalleux-ludo) [reply scripts/create-price-discovery-offer.ts:39]: Changed. Now reading the WETH address from the coreSDK config

File: packages/core-sdk/src/seaport/interface.ts (hunk 5 above)
Comment (levalleux-ludo) [reply packages/core-sdk/src/seaport/interface.ts:16]: These values are the ones used in the smart contract. So no way to change them

File: packages/common/src/types/offers.ts (hunk 4 above)
Comment (levalleux-ludo) [reply packages/common/src/types/offers.ts:80]: These values are the ones used in smart contract.
Creating a shift in coreSDK compared to the smart contract would raise confusion...

File: e2e/opensea-api-mock/src/utils/logger.ts (hunk 2 above)
Comment (levalleux-ludo) [reply e2e/opensea-api-mock/src/utils/logger.ts:1]: Interesting question. 
Same code as in meta-tx-gateway.
I'd say morgan is just used to format the HTTP requests, seems ideal for that.
While Winston is used to log out globally (not only at HTTP level).
Morgan redirect to Winston anyway, so it doesn't really appear to be redundant.

File: packages/core-sdk/src/price-discovery/handler.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+}): Promise<TransactionResponse> {
+  let { offerId } = exchanges.parseTokenId(args.tokenIdOrOfferId);
+  if (offerId.eq(0)) {
+    offerId = BigNumber.from(args.tokenIdOrOfferId);
+  }
+  const offer = await getOfferById(args.subgraphUrl, offerId);
+
+  // is Offer committable condition is different from classic
+  //  usecase (for instance quantityAvailable can be 0, as the rNFT are already preminted)
+  // await exchanges.checkOfferIsCommittable(offerId, offer);
```
[the remark this answers is not in the document]
Comment (levalleux-ludo) [reply packages/core-sdk/src/price-discovery/handler.ts:28]: ok

File: e2e/opensea-api-mock/src/services/opensea.ts (hunk 1 above)
Comment (levalleux-ludo) [reply e2e/opensea-api-mock/src/services/opensea.ts:172]: ok

**Pull request merged**